[lldb] Revive internal data formatter documentation (#66527)
llvm/llvm-project: lldb/docs/resources/dataformatters.rst

Data Formatters
===============

This page is an introduction to the design of the LLDB data formatters
subsystem. The intended target audience are people interested in understanding
or modifying the formatters themselves rather than writing a specific data
formatter. For this latter purpose, the user documentation about formatters is
the main relevant document which one should refer to.

This page also highlights some open areas for improvement to the general
subsystem, and more evolutions not anticipated here are certainly possible.

Overview
--------

The LLDB data formatters subsystem is used to allow the debugger as well as the
end-users to customize the way their variables look upon inspection in the user
interface (be it the command line tool, or one of the several GUIs that are

> backed by LLDB)

Reviewer: ```suggestion
backed by LLDB).
```

To this aim, they are hooked into the ValueObjects model, in order to provide
entry points through which such customization questions can be answered. For

> example what format should this number be printed as? How many child elements

Reviewer: ```suggestion
example: What format should this number be printed as? How many child elements
```

formatter types are backed by a class hierarchy.

Summaries, for instance, can exist in one of three "flavors":

* Summary strings
* Python script
* Native C++

The base class for summaries, ``TypeSummaryImpl``, is a pure virtual class that
wraps, again, the Flags, and exports among others:

::

  virtual bool FormatObject (ValueObject *valobj, std::string& dest) = 0;

This is the core entry point, which allows subclasses to specify their mode of
operation.

``StringSummaryFormat``, which is the class that implements summary strings,
does a check as to whether the summary is a one-liner, and if not, then uses
its stored summary string to call into ``Debugger::FormatPrompt``, and obtain a
string back, which it returns in ``dest`` as the resulting summary.

For a Python summary, implemented in ``ScriptSummaryFormat``,
``FormatObject()`` calls into the ``ScriptInterpreter`` which is supposed to
hold the knowledge on how to bridge back and forth with the scripting language
(Python in the case of LLDB) in order to produce a valid string. Implementors
of new ``ScriptInterpreters`` for other languages are expected to provide a

> ``GetScriptedSummary()`` entrypoint for this purpose, if they desire to allow

Reviewer: ```suggestion
``GetScriptedSummary()`` entry point for this purpose, if they desire to allow
```

and call into it to execute their duty. It should be noted that there are no
facilities for users to interact with C++ formatters, and as such they are
extremely opaque, effectively being a thin wrapper between plain function
pointers and the LLDB formatters subsystem.

Also, dynamic loading of C++ formatters in LLDB is currently not implemented,
and as such it is safe and reasonable for these formatters to deal with
internal ``ValueObjects`` instances instead of public ``SBValue`` objects.

An interesting data point is that summaries are expected to be stateless. While
at the Python layer they are handed an ``SBValue`` (since nothing else could be
visible for scripts), it is not expected that the ``SBValue`` should be cached
and reused - any and all caching occurs on the LLDB side, completely
transparent to the formatter itself.

The design of synthetic children is somewhat more intricate, due to them being
stateful objects. The core idea of the design is that synthetic children act
like a two-tier model, in which there is a backend dataset (the underlying
unformatted ``ValueObject``), and an higher level view (frontend) which vends

> the computed representation

Reviewer: ```suggestion
the computed representation.
```

number of children, and then for each of them as necessary, they should be
prepared to return a ``ValueObject`` describing the child. They might also be
asked to provide a name-to-index mapping (e.g. to allow LLDB to resolve queries
like ``myFoo.myChild``).

``Update()`` and ``MightHaveChildren()`` are described in the user
documentation, and they mostly serve bookkeeping purposes.

LLDB provides three kinds of synthetic children: filters, scripted synthetics,
and the native C++ providers Filters are implemented by
``TypeFilterImpl::FrontEnd``.

Scripted synthetics are implemented by ``ScriptedSyntheticChildren::FrontEnd``,
plus a set of callbacks provided by the ``ScriptInterpteter`` infrastructure to
allow LLDB to pass the front-end queries down to the scripting languages.

As for C++ native synthetics, there is a ``CXXSyntheticChildren``, but no
corresponding ``FrontEnd`` class. The reason for this design is that
``CXXSyntheticChildren`` store a callback to a creator function, which is
responsible for providing a ``FrontEnd``. Each individual formatter (e.g.
``LibstdcppMapIteratorSyntheticFrontEnd``) is a standalone frontend, and once

> created retains to relation to its underlying ``SyntheticChildren`` object

Reviewer: ```suggestion
created retains to relation to its underlying ``SyntheticChildren`` object.
```

On a ``ValueObject`` level, upon being asked to generate synthetic children for
a ``ValueObject``, LLDB spawns a ValueObjectSynthetic object which is a
subclass of ``ValueObject``. Building upon the ``ValueObject`` infrastructure,
it stores a backend, and a shared pointer to the ``SyntheticChildren``. Upon
being asked queries about children, it will use the ``SyntheticChildren`` to
generate a front-end for itself and will let the front-end answer questions.
The reason for not storing the ``FrontEnd`` itself is that there is no
guarantee that across updates, the same ``FrontEnd`` will be used over and over
(e.g. a ``SyntheticChildren`` object could serve an entire class hierarchy and

> vend different frontends for different subclasses)

Reviewer: ```suggestion
vend different frontends for different subclasses).
```
